### ISSUE TITLE
Feature Request: Support for Multiple Sources in Map Styles Using ol-maplibre-layer

### ISSUE THREAD

**Issue by asebert-caue** [IGNF-Ma-carte/mcutils#1]
Hello,

We commonly use Maputnik to create **Maplibre styles** ([maplibre-style-spec](https://maplibre.org/maplibre-style-spec/)) that combine **multiple sources**.

We have noticed that MaCarte uses the **ol-mapbox-style** library. After testing, we discovered that it can only load a **single source**. As stated in the documentation: _"The style function will render all layers from the glStyle object that use the source of the first layer"_ ([ol-mapbox-style - applyStyle](https://github.com/openlayers/ol-mapbox-style?tab=readme-ov-file#applyStyle)).

However, we tested the ol-maplibre-layer library ([ol-maplibre-layer](https://github.com/geoblocks/ol-maplibre-layer/tree/master?tab=readme-ov-file)). We were able to create multiple layers by loading Maplibre styles containing multiple sources into a map.

MapLibre is an open-source library, unlike Mapbox. Therefore, it might be more appropriate to use it for rendering style files, which aligns more with the open-source logic promoted by IGN.

Thank you for your work.

**Comment by Viglino** [comment]
Hello,

I added the feature to handle multiple layers in vector tiles (using ol-maplibre-layer).
It seems to work.
NB1: only works for local files (no remote files)
NB2: the layer is no longer an ol layer and does not benefit from the selection/hover popup
NB3: an error message in the console when rendering on printing (to report)

Online test: https://ignf-ma-carte.github.io/mcutils/mcutils/charte.html?map=FfyxzB

@+

**Comment by asebert-caue** [comment]
It's good news that it works.

NB1: only works for local files (no remote files)

Do you have more details about this issue? We have tested this library with an online style and did not encounter this problem.
Could it be a CORS issue?

Here is a link to a test style that mixes several sources:
[https://tileserver.s-pass.org/styles/topo4colorsworld/style.json](https://tileserver.s-pass.org/styles/topo4colorsworld/style.json)

And another link to visualize the rendering on MapLibre:
[https://tileserver.s-pass.org/styles/topo4colorsworld/?vector#14.57/51.00246/2.06029](https://tileserver.s-pass.org/styles/topo4colorsworld/?vector#14.57/51.00246/2.06029)

NB2: the layer is no longer an ol layer and does not benefit from the selection/hover popup

I understand, we are in a particular case reserved for experts. It might be possible to manage both types.
When you create styles with Maputnik, it is possible to choose the rendering mode MapLibre GL JS or OpenLayers. 

![CleanShot 2024-07-30 at 10 34 59@2x](https://github.com/user-attachments/assets/432028d1-2e8d-4ae1-a384-fdd86c95eac0)

This choice is then recorded in the [Metadata property of the style](https://maplibre.org/maplibre-style-spec/root/#metadata):
```json
{
  "version": 8,
  "name": "SPASS",
  "metadata": {
    "mapbox:autocomposite": false,
    "mapbox:type": "template",
    "maputnik:renderer": "mbgljs", // or "ol"
    "openmaptiles:version": "3.x",
    "openmaptiles:mapbox:owner": "openmaptiles",
    "openmaptiles:mapbox:source:url": "mapbox://openmaptiles.4qljc88t"
  }
}
```

If this information is not present, you could use the rendering mode with ol-mapbox to maintain the initial behavior of the viewer and thus retain the selection/hover and popup functionalities.
You can also rename this property, for example to `mc:renderer`.
The advantages and disadvantages of each rendering mode could be indicated in the documentation.

NB3: an error message in the console when rendering on printing (to report)

Indeed, the printing functionality can be problematic. This is why implementing both rendering modes could be a solution, with proper documentation.

**Comment by Viglino** [comment]
> NB1: only works for local files (no remote files)

Not a real issue: the caller has to know the layer type to use  and switch between ol or Maplibre layer.
For now, it checks the number of layers and switches to Maplibre if more than one (that's why we need the local file). We have to add a parameter to set the layer type.

> NB2: the layer is no longer an ol layer and does not benefit from the selection/hover popup

We have to add a hack to query the Maplibre map on click (but not implemented yet).

> NB3: an error message in the console when rendering on printing

Not a real issue, the printing works but an error is fired on first attempt to render the map (but it finally renders).

**Comment by asebert-caue** [comment]
Hello,

Thank you again for your responsiveness. I noticed that the feature is implemented in the main branch of mcutils and that version 3.4.5 of the mcviewer uses this version of mcutils. However, this version of the viewer is not yet in production, is that right?

Have a nice day.

**Comment by Viglino** [comment]
Not yet! 
v.3.4 is still in the testing phase and will be available at the end of September. 
You can test it on the qualification site: https://qlf-macarte.ign.fr/

**Comment by asebert-caue** [comment]
Thank you for the information. I will test it and provide detailed feedback.

**Comment by asebert-caue** [comment]
Hello,

I tested version 3.4 on the qualification site, but I don't have the impression that styles with multiple sources work. I tried with a remote file and a local file.

Here are two different styles : 

**Topo 4 couleurs Monde**
**Style** : https://api.s-pass.org/mapstyles/f9c05c82-e645-40fd-9dc9-fffd075328bd.json
**Preview** : https://maputnik.s-pass.org?style=https://api.s-pass.org/mapstyles/f9c05c82-e645-40fd-9dc9-fffd075328bd.json

**Satellite Hybride France**
**Style** : https://api.s-pass.org/mapstyles/d3b27b30-e99f-4e80-a92a-63729d783731.json
**Preview** : https://maputnik.s-pass.org?style=https://api.s-pass.org/mapstyles/d3b27b30-e99f-4e80-a92a-63729d783731.json

**Comment by Viglino** [comment]
It works for me:
- https://qlf-macarte.ign.fr/carte/c3Sy5j/Topo-4-couleurs-Monde-MVT
- https://qlf-macarte.ign.fr/carte/uWtEJo/Satellite-Hybride-France-MVT

NB1: only with a local file
NB2: not completly integrated in the UI (when you first set the layer it doesn't use the multilayer but when you reload it works)

PS: I'll try to fix 2 and a bug on writing/loading.

**Comment by Viglino** [comment]
Just push a new commit to fix NB2 (https://github.com/IGNF-Ma-carte/mceditor/commit/b6213297d19232f35d7ee2e041ca19ca8834062d).
Deploy to qlf asap.

**Comment by asebert-caue** [comment]
that's right, it works when you reload the map! That's great!! Good job. 🚀🚀

**Comment by Viglino** [comment]
v4.3.6 is online in qlf: https://qlf-macarte.ign.fr/edition/carte/
Now you can see / save multi source MVT (local file).
